File: packages/observability/src/metrics.ts
export interface Counter {
  inc(labels?: Record<string, string>, delta?: number): void;
  get(labels?: Record<string, string>): number;
}

export interface Gauge {
  set(value: number, labels?: Record<string, string>): void;
  get(labels?: Record<string, string>): number;
}

export interface HistogramSnapshot {
  count: number;
  sum: number;
  p50: number;
  p95: number;
  p99: number;
  min: number;
  max: number;
}

export interface Histogram {
  observe(value: number, labels?: Record<string, string>): void;
  snapshot(labels?: Record<string, string>): HistogramSnapshot;
}

/** Max observations per histogram series to bound memory. */
const MAX_OBSERVATIONS = 10_000;

/**
 * In-memory metrics registry.
 * Counters and histograms keyed by name + serialized labels.
 */
export class MetricsRegistry {
  private counters = new Map<string, Map<string, number>>();
  private gauges = new Map<string, Map<string, number>>();
  private histograms = new Map<string, Map<string, number[]>>();

  createGauge(name: string): Gauge {
    if (!this.gauges.has(name)) {
      this.gauges.set(name, new Map());
    }
    const store = this.gauges.get(name)!;

    return {
      set: (value, labels = {}) => {
        store.set(labelsKey(labels), value);
      },
      get: (labels = {}) => store.get(labelsKey(labels)) ?? 0,
    };
  }

  createCounter(name: string): Counter {
    if (!this.counters.has(name)) {
      this.counters.set(name, new Map());
    }
    const store = this.counters.get(name)!;

    return {
      inc: (labels = {}, delta = 1) => {
        const key = labelsKey(labels);
        store.set(key, (store.get(key) ?? 0) + delta);
      },
      get: (labels = {}) => store.get(labelsKey(labels)) ?? 0,
    };
  }

  createHistogram(name: string): Histogram {
    if (!this.histograms.has(name)) {
      this.histograms.set(name, new Map());
    }
    const store = this.histograms.get(name)!;

    return {
      observe: (value, labels = {}) => {
        const key = labelsKey(labels);
        if (!store.has(key)) store.set(key, []);
        const arr = store.get(key)!;
        arr.push(value);
        if (arr.length > MAX_OBSERVATIONS) {
          arr.splice(0, arr.length - MAX_OBSERVATIONS);
        }
      },
      snapshot: (labels = {}) => {
        const arr = store.get(labelsKey(labels)) ?? [];
        return computeSnapshot(arr);
      },
    };
  }

  /** Serialize all metrics as a flat JSON object for /metrics endpoint. */
  toJSON(): Record<string, unknown> {
    const result: Record<string, unknown> = {};

    for (const [name, store] of this.counters) {
      for (const [key, value] of store) {
        result[key ? `${name}{${key}}` : name] = value;
      }
    }

    for (const [name, store] of this.gauges) {
      for (const [key, value] of store) {
        result[key ? `${name}{${key}}` : name] = value;
      }
    }

    for (const [name, store] of this.histograms) {
      for (const [key, arr] of store) {
        const prefix = key ? `${name}{${key}}` : name;
        const snap = computeSnapshot(arr);
        result[`${prefix}_count`] = snap.count;
        result[`${prefix}_sum`] = round(snap.sum);
        result[`${prefix}_p50`] = round(snap.p50);
        result[`${prefix}_p95`] = round(snap.p95);
        result[`${prefix}_p99`] = round(snap.p99);
      }
    }

    return result;
  }

  /** Reset all metrics (for testing). */
  reset(): void {
    this.counters.clear();
    this.gauges.clear();
    this.histograms.clear();
  }
}

function labelsKey(labels: Record<string, string>): string {
  const keys = Object.keys(labels).sort();
  if (keys.length === 0) return "";
  return keys.map((k) => `${k}="${labels[k]}"`).join(",");
}

function computeSnapshot(arr: number[]): HistogramSnapshot {
  if (arr.length === 0) {
    return { count: 0, sum: 0, p50: 0, p95: 0, p99: 0, min: 0, max: 0 };
  }
  const sorted = [...arr].sort((a, b) => a - b);
  return {
    count: sorted.length,
    sum: sorted.reduce((s, v) => s + v, 0),
    p50: percentile(sorted, 0.5),
    p95: percentile(sorted, 0.95),
    p99: percentile(sorted, 0.99),
    min: sorted[0],
    max: sorted[sorted.length - 1],
  };
}

function percentile(sorted: number[], p: number): number {
  const idx = Math.ceil(p * sorted.length) - 1;
  return sorted[Math.max(0, idx)];
}

function round(n: number): number {
  return parseFloat(n.toFixed(2));
}

File: apps/embedding-worker/src/metrics.py
"""In-memory metrics for the embedding worker.

Mirrors the TypeScript MetricsRegistry pattern — counters and histograms
with a to_json() method for the /metrics HTTP endpoint.
"""

import json
import math
import threading
from typing import Any


class Counter:
    def __init__(self):
        self._lock = threading.Lock()
        self._value = 0

    def inc(self, delta: int = 1) -> None:
        with self._lock:
            self._value += delta

    def get(self) -> int:
        return self._value


class Histogram:
    MAX_OBSERVATIONS = 10_000

    def __init__(self):
        self._lock = threading.Lock()
        self._values: list[float] = []

    def observe(self, value: float) -> None:
        with self._lock:
            self._values.append(value)
            if len(self._values) > self.MAX_OBSERVATIONS:
                self._values = self._values[-self.MAX_OBSERVATIONS:]

    def snapshot(self) -> dict[str, float]:
        with self._lock:
            arr = sorted(self._values)
        if not arr:
            return {"count": 0, "sum": 0, "p50": 0, "p95": 0, "p99": 0, "min": 0, "max": 0}
        return {
            "count": len(arr),
            "sum": round(sum(arr), 2),
            "p50": _percentile(arr, 0.50),
            "p95": _percentile(arr, 0.95),
            "p99": _percentile(arr, 0.99),
            "min": arr[0],
            "max": arr[-1],
        }


class Gauge:
    def __init__(self):
        self._lock = threading.Lock()
        self._value: float = 0

    def set(self, value: float) -> None:
        with self._lock:
            self._value = value

    def get(self) -> float:
        return self._value


def _percentile(sorted_arr: list[float], p: float) -> float:
    idx = max(0, math.ceil(p * len(sorted_arr)) - 1)
    return round(sorted_arr[idx], 2)


class MetricsRegistry:
    def __init__(self):
        self._counters: dict[str, Counter] = {}
        self._gauges: dict[str, Gauge] = {}
        self._histograms: dict[str, Histogram] = {}

    def create_counter(self, name: str) -> Counter:
        c = Counter()
        self._counters[name] = c
        return c

    def create_gauge(self, name: str) -> Gauge:
        g = Gauge()
        self._gauges[name] = g
        return g

    def create_histogram(self, name: str) -> Histogram:
        h = Histogram()
        self._histograms[name] = h
        return h

    def to_json(self) -> dict[str, Any]:
        result: dict[str, Any] = {}
        for name, c in self._counters.items():
            result[name] = c.get()
        for name, g in self._gauges.items():
            result[name] = g.get()
        for name, h in self._histograms.items():
            snap = h.snapshot()
            result[f"{name}_count"] = snap["count"]
            result[f"{name}_sum"] = snap["sum"]
            result[f"{name}_p50"] = snap["p50"]
            result[f"{name}_p95"] = snap["p95"]
            result[f"{name}_p99"] = snap["p99"]
        return result


# ── Singleton registry and metrics ─────────────────────────────

registry = MetricsRegistry()

embedding_jobs_total = registry.create_counter("embedding_jobs_total")
embedding_job_failures_total = registry.create_counter("embedding_job_failures_total")
embedding_latency_ms = registry.create_histogram("embedding_latency_ms")
embeddings_inserted_total = registry.create_counter("embeddings_inserted_total")
job_retry_total = registry.create_counter("embedding_job_retry_total")
job_dlq_total = registry.create_counter("embedding_job_dlq_total")
job_poison_total = registry.create_counter("embedding_job_poison_total")
queue_depth = registry.create_gauge("embedding_queue_depth")
dlq_depth = registry.create_gauge("embedding_dlq_depth")
